pkg/metrics/metrics.go

```diff
@@ -0,0 +1,98 @@
+package metrics
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Error to be  returned when marketplace is unable to reach quay.io.
+type OpsrcError string
+
+const (
+	// Default values which are to be used as labels
+	// for prometheus metrics.
+	Error     = "opsrc_error"
+	OpsrcName = "opsrc_name"
+
+	// Path, Port and Host where custom metrics would be exposed.
+	MetricsPath = "/metrics"
+	MetricsPort = "8383"
+	MetricsHost = "0.0.0.0"
+
+	// Error message to be returned when quay.io is down.
+	UnreachableError OpsrcError = "quay_unavailabe"
+
+	// Error message to be returned when host address is wrong.
+	NoHostExsistsError OpsrcError = "no_host_exsists"
+)
+
+// Metric variables that report error based on the operator-source
+// which fails to reconcile.
+var (
+	operatorSourceError = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "operator_source_failure_count",
+			Help: "Monotonic count of times operator-source could not download packages",
+		},
+		[]string{OpsrcName, Error},
+	)
+	// List of custom metrics which are to be collected.
+	metricsList = []prometheus.Collector{
+		operatorSourceError,
+	}
+)
+
+// registerMetrics registers the metrics with prometheus.
+func registerMetrics() error {
+	for _, metric := range metricsList {
+		err := prometheus.Register(metric)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RecordMetrics starts the server, records the custom metrics and exposes
+// the metrics at the specified endpoint.
+func RecordMetrics() {
+	// Register metrics for the operator with the prometheus.
+	registerMetrics()
+
+	// Start the server and expose the registered metrics.
+	http.Handle(MetricsPath, promhttp.Handler())
+	port := ":" + MetricsPort
+	go http.ListenAndServe(port, nil)
+}
+
+// counterForOpsrcError returns the number of times error was encountered while
+// contacting quay for operstor-sources.
+func counterForOpsrcError(opsrc, opsrcError string) prometheus.Counter {
+	return operatorSourceError.WithLabelValues(opsrc, opsrcError)
+}
+
+// updateMetrics takes the name of the operator source which fails to reconcile, and
+// increments the error count of the respective operator metric.
+func UpdateMetrics(opsrc, opsrcError string) {
+	counterForOpsrcError(opsrc, opsrcError).Inc()
+}
+
+// ErrorMessage takes the error which appears while Listing the packages,
+// parses it, and returns the error message which is to be published.
+func ErrorMessage(err string) OpsrcError {
```

Comment:
This is a good start, we will need to review how we wish to ID quay errors.

Reply:
Agreed. It seems like string comparison will be brittle (if these error messages change, this code will silently break).

Perhaps we can derive meaning from control flow instead? Or at least http error codes if we can get them from the client?

Reply:
Yeah, true. We do get a http 404 error code when quay is not available. But when the endpoint is wrong, it remains in `configuring` phase, trying to connect. There is no specific error code returned in the error message. I am not sure on how to proceed with that case though..

Reply:
We put ourselves back in the configuring phase because we know there was some error and we want to retry. We should be able to dig into where we decide to put ourselves back into that phase and probably learn more. And, if there isn't something more specific being returned, is it possible for us to adjust how the appr client is giving a result? Are we really restricted to just strings? Could we update the error that is returned to have an `ErrorCode` field or something similar?